### PR TITLE
add cloudy flag to ghi and nighttime to power variables

### LINKED ISSUE #210: add detect_clearsky_ghi to ghi validation

### LINKED ISSUE #293: Add NIGHTTIME validation for power variables

### PR DESCRIPTION
**Pull request by alorenzo175**
- [x] Closes #210 closes #293 .
  - [x] I am familiar with the [contributing guidelines](https://solarforecastarbiter-core.readthedocs.io/en/latest/contributing.html).
  - [x] Tests added.
  - [x] Updates entries to [`docs/source/api.rst`](https://github.com/SolarArbiter/solarforecastarbiter-core/blob/master/docs/source/api.rst) for API changes.
  - [x] Adds descriptions to appropriate "what's new" file in [`docs/source/whatsnew`](https://github.com/SolarArbiter/solarforecastarbiter-core/tree/master/docs/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
  - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
  - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

### CODE REVIEW

**Comment by alorenzo175** [comment]
I've modified the validation.validator.detect_clearsky_ghi to raise warnings instead of errors and default to flagging a time as cloudy if the pvlib detect clearsky can't run. This produces a validation like
image

I think this is probably the right approach, but wanted to get some feedback @wholmgren @cwhanse

File: solarforecastarbiter/validation/tasks.py [hunk 1]
```diff
@@ -60,7 +62,10 @@ def validate_ghi(observation, values):
         _return_mask=True)
     ghi_clearsky_flag = validator.check_ghi_clearsky(values, clearsky['ghi'],
                                                      _return_mask=True)
-    return timestamp_flag, night_flag, ghi_limit_flag, ghi_clearsky_flag
+    cloudy_flag = validator.detect_clearsky_ghi(values, clearsky['ghi'],
+                                                _return_mask=True)
```
Comment (wholmgren) [inline solarforecastarbiter/validation/tasks.py:66]: I thought you'd want a try/except here instead of a warning in the function.

**Comment by alorenzo175** [comment]
> The results look good to me. When does the pvlib function fail with our data?

It will fail with too little data or uneven timestamps or long interval lengths

File: solarforecastarbiter/validation/tasks.py (hunk 1 above)
Comment (alorenzo175) [reply solarforecastarbiter/validation/tasks.py:66]: I went back and forth on this. That is one option, but the rest of the functions in validator.py will not raise an error unless an invalid parameter for the algorith is chosen vs invalid data. 

Also, it's much easier to return a series of 0 in the function and have those converted to the flag vs having to pull in logic from quality_mapping.py to figure out what the flag should be in the except

**Comment by cwhanse** [comment]
> It will fail with too little data or uneven timestamps or long interval lengths

Or missing timestamps. pvlib #[507](https://github.com/pvlib/pvlib-python/issues/507#issue-342341823)

**Comment by cwhanse** [comment]
> I've modified the validation.validator.detect_clearsky_ghi to raise warnings instead of errors and default to flagging a time as cloudy if the pvlib detect clearsky can't run. This produces a validation like
> 
> I think this is probably the right approach, but wanted to get some feedback @wholmgren @cwhanse

The label "cloudy" could be misleading, as it's either not clear, or indeterminate. Maybe reverse that part of the plot and color times labeled as Clear?

I do think that most of the validation shouldn't apply to night times.

**Comment by alorenzo175** [comment]
> The label "cloudy" could be misleading, as it's either not clear, or indeterminate. Maybe reverse that part of the plot and color times labeled as Clear?

An added benefit of inverting to a Clear or Cloud Free flag is that the default behavior of excluding positive quality flags from reports for these periods makes more sense. I.e. exclude the periods of clearsky from a GHI report vs excluding the cloudy periods.

**Comment by alorenzo175** [comment]
Ok, I've switched to Cloud Free and only flagging times that the detect clearsky algorithm flags as clear.

image

We can change from CLOUD FREE, but I thought it was more clear when we already have CLEARSKY EXCEEDED

**Comment by alorenzo175** [comment]
fair, CLEAR is too ambiguous, but CLEARSKY is ok

File: solarforecastarbiter/validation/tasks.py [hunk 2]
```diff
@@ -47,7 +52,8 @@ def validate_ghi(observation, values):
         `validator.check_timestamp_spacing`,
         `validator.check_irradiance_day_night`,
         `validator.check_ghi_limits_QCRad`,
-        `validator.check_ghi_clearsky`
+        `validator.check_ghi_clearsky`,
+        `validator.detect_clearsky_ghi`
```
Comment (wholmgren) [inline solarforecastarbiter/validation/tasks.py:56]: not a big deal, but I believe in numpydoc format the tuple is implied by multiple return values, so proper format would be

```
Returns
-------
timestamp_flag
    bitmask from `validator.check_timestamp_spacing`
night_flag
    bitmask from `validator.check_irradiance_day_night`

etc
```

File: solarforecastarbiter/validation/validator.py [hunk 3]
```diff
@@ -706,10 +705,13 @@ def detect_clearsky_ghi(ghi, ghi_clearsky):
     Clear-Sky Periods From Irradiance Data," in IEEE Journal of Photovoltaics,
     vol. 9, no. 4, pp. 998-1005, July 2019. doi: 10.1109/JPHOTOV.2019.2914444
     """
+    if len(ghi) < 2:
+        return pd.Series(0, index=ghi.index)
```
Comment (wholmgren) [inline solarforecastarbiter/validation/validator.py:709]: `RuntimeWarning` here too?

**Pull request merged**